cmd/xsDeploy.go

```diff
@@ -0,0 +1,424 @@
+package cmd
+
+import (
+	"bytes"
+	"fmt"
+	"github.com/SAP/jenkins-library/pkg/command"
+	"github.com/SAP/jenkins-library/pkg/log"
+	"github.com/SAP/jenkins-library/pkg/piperutils"
+	"github.com/pkg/errors"
+	"io"
+	"os"
+	"regexp"
+	"strings"
+	"sync"
+	"text/template"
+)
+
+//
+// START DeployMode
+type DeployMode int
+
+const (
+	// NoDeploy ...
+	NoDeploy DeployMode = iota
+	//Deploy ...
+	Deploy DeployMode = iota
+	//BGDeploy ...
+	BGDeploy DeployMode = iota
+)
+
+//ValueOfMode ...
+func ValueOfMode(str string) (DeployMode, error) {
+	switch str {
+	case "NONE":
+		return NoDeploy, nil
+	case "DEPLOY":
+		return Deploy, nil
+	case "BG_DEPLOY":
+		return BGDeploy, nil
+	default:
+		return NoDeploy, errors.New(fmt.Sprintf("Unknown DeployMode: '%s'", str))
+	}
+}
+
+// String
+func (m DeployMode) String() string {
+	return [...]string{
+		"NONE",
+		"DEPLOY",
+		"BG_DEPLOY",
+	}[m]
+}
+
+// END DeployMode
+//
+
+//
+// START Action
+type Action int
+
+const (
+	//None ...
+	None Action = iota
+	//Resume ...
+	Resume Action = iota
+	//Abort ...
+	Abort Action = iota
+	//Retry ...
+	Retry Action = iota
+)
+
+//ValueOfAction ...
+func ValueOfAction(str string) (Action, error) {
+	switch str {
+	case "NONE":
+		return None, nil
+	case "RESUME":
+		return Resume, nil
+	case "ABORT":
+		return Abort, nil
+	case "RETRY":
+		return Retry, nil
+
+	default:
+		return None, errors.New(fmt.Sprintf("Unknown Action: '%s'", str))
+	}
+}
+
+// String
+func (a Action) String() string {
+	return [...]string{
+		"NONE",
+		"RESUME",
+		"ABORT",
+		"RETRY",
+	}[a]
+}
+
+// END Action
+//
+
+const loginScript = `#!/bin/bash
+xs login -a {{.APIURL}} -u {{.User}} -p '{{.Password}}' -o {{.Org}} -s {{.Space}} {{.LoginOpts}}
+`
+
+const logoutScript = `#!/bin/bash
+xs logout`
+
+const deployScript = `#!/bin/bash
+xs {{.Mode}} {{.MtaPath}} {{.DeployOpts}}`
+
+const completeScript = `#!/bin/bash
+xs {{.Mode.GetDeployCommand}} -i {{.DeploymentID}} -a {{.Action.GetAction}}
+`
+
+func xsDeploy(myXsDeployOptions xsDeployOptions) error {
+	c := command.Command{}
+	return runXsDeploy(myXsDeployOptions, &c, piperutils.FileExists, piperutils.Copy, os.Remove)
+}
+
+func runXsDeploy(XsDeployOptions xsDeployOptions, s shellRunner,
+	fExists func(string) bool,
+	fCopy func(string, string) (int64, error),
+	fRemove func(string) error) error {
+
+	mode, err := ValueOfMode(XsDeployOptions.Mode)
+	if err != nil {
+		fmt.Printf("Extracting mode failed: %v\n", err)
```

Comment:
do not use `fmt.Print...` but rather `log.Entry().WithError()`

Reply:
I fully agree

Reply:
fixed, replaced by proper exception wrapping